File: test/test_board.rb
require 'test/ruby_tanbo_test.rb'
require 'main_controller.rb'

class BoardTest < Test::Unit::TestCase
  include RubyTanboTest

  def test_start_position
    msg = "Starting position was found invalid, wrong moves for WHITE at: "
    point = @main_controller.get_board[6, 0]
    assert_equal TanboBoard::WHITE, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[18, 0]
    assert_equal TanboBoard::WHITE, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[0, 6]
    assert_equal TanboBoard::WHITE, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[12, 6]
    assert_equal TanboBoard::WHITE, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[6, 12]
    assert_equal TanboBoard::WHITE, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[18, 12]
    assert_equal TanboBoard::WHITE, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[0, 18]
    assert_equal TanboBoard::WHITE, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[12, 18]
    assert_equal TanboBoard::WHITE, @main_controller.get_color(point), (msg + point.inspect)
               
    msg = "Starting position was found invalid, wrong moves for BLACK at: "
    point = @main_controller.get_board[0, 0]
    assert_equal TanboBoard::BLACK, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[12, 0]
    assert_equal TanboBoard::BLACK, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[6, 6]
    assert_equal TanboBoard::BLACK, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[18, 6]
    assert_equal TanboBoard::BLACK, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[12, 12]
    assert_equal TanboBoard::BLACK, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[18, 18]
    assert_equal TanboBoard::BLACK, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[6, 18]
    assert_equal TanboBoard::BLACK, @main_controller.get_color(point), (msg + point.inspect)
    point = @main_controller.get_board[0, 12]
    assert_equal TanboBoard::BLACK, @main_controller.get_color(point), (msg + point.inspect)
  
    msg = "Starting position was found invalid, expected EMPTY point not empty at: "
    0.upto(18) do |x|
      0.upto(18) do |y|
        unless x%6 == 0 && y%6 == 0
          point = @main_controller.get_board[x, y]
          assert_equal TanboBoard::BLANK, @main_controller.get_color(point), (msg + point.inspect)
        end
      end
    end
  end

  def test_black_starts
    assert_equal TanboBoard::BLACK, @main_controller.whose_turn?
  end

  def test_valid_move
    msg = "Valid WHITE starting move reported as invalid at: "
    point = @main_controller.get_board[0, 7]
    assert @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[0, 5]
    assert @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[1, 18]
    assert @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[0, 17]
    assert @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[5, 12]
    assert @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[6, 13]
    assert @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
  
    msg = "Valid BLACK starting move reported as invalid at: "
    point = @main_controller.get_board[0, 1]
    assert @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    point = @main_controller.get_board[1, 0]
    assert @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    point = @main_controller.get_board[17, 18]
    assert @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    point = @main_controller.get_board[18, 17]
    assert @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    point = @main_controller.get_board[7, 6]
    assert @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    point = @main_controller.get_board[6, 7]
    assert @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
  end

  def test_no_invalid_move
    msg = "Invalid move from starting position reported as valid at: "
    point = @main_controller.get_board[-1,-1]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[19,19]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[-1,19]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[19,-1]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[-532,1087]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[7734,-1985]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[3,4]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[4,4]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[5,4]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[9,0]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[0,10]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[10,0]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[0,18]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[6,6]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[10,18]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
    point = @main_controller.get_board[18,10]
    assert ! @main_controller.valid_move?(point, TanboBoard::BLACK), (msg + point.inspect)
    assert ! @main_controller.valid_move?(point, TanboBoard::WHITE), (msg + point.inspect)
  end
  
  def test_right_root_number
    assert_equal 16, @main_controller.roots.size
  end
  
  def test_root_points
    root_points = [
      @main_controller.get_board[6 ,0 ], @main_controller.get_board[18,0 ],
      @main_controller.get_board[0 ,6 ], @main_controller.get_board[12,6 ],
      @main_controller.get_board[6 ,12], @main_controller.get_board[18,12],
      @main_controller.get_board[0 ,18], @main_controller.get_board[12,18],
      @main_controller.get_board[0 ,0 ], @main_controller.get_board[12,0 ],
      @main_controller.get_board[6 ,6 ], @main_controller.get_board[18,6 ],
      @main_controller.get_board[12,12], @main_controller.get_board[18,18],
      @main_controller.get_board[6 ,18], @main_controller.get_board[0 ,12]
    ]
    
    for point in root_points
      found = []
      for root in @main_controller.roots
        deleted = root.points.delete(point)
        found << deleted if deleted
      end
      
      assert !found.empty?, "Couldn't find point in any root: " + point.inspect
      assert_equal 1, found.size, "More than one root had point: " + point.inspect
    end
  end
  
  def test_root_liberties
    root_points = [
      @main_controller.get_board[6 ,0 ], @main_controller.get_board[18,0 ],
      @main_controller.get_board[0 ,6 ], @main_controller.get_board[12,6 ],
      @main_controller.get_board[6 ,12], @main_controller.get_board[18,12],
      @main_controller.get_board[0 ,18], @main_controller.get_board[12,18],
      @main_controller.get_board[0 ,0 ], @main_controller.get_board[12,0 ],
      @main_controller.get_board[6 ,6 ], @main_controller.get_board[18,6 ],
      @main_controller.get_board[12,12], @main_controller.get_board[18,18],
      @main_controller.get_board[6 ,18], @main_controller.get_board[0 ,12]
    ]
    
    for point in root_points
      for root in @main_controller.roots
        if root.points.include?(point)
          point.bounded_neighbors.each do |lib|
            assert root.liberties.include?(lib), "Root: " + root.inspect + " should have had liberty: " + lib.inspect
          end
        end
      end
    end
  end
  
end

File: ai/ruby_uct/test/test_board.cpp
// Author: Travis Briggs, briggs.travis (at) gmail.com
// ===================================================
// Copyright (C) 2009 Travis Briggs
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. See the COPYING file. If not, see
// <http://www.gnu.org/licenses/>.

#include "common.h"
#include "test_board.h"

#include "point_tanbo.h"
#include "root_tanbo.h"

#include <boost/shared_ptr.hpp>

#include <vector>
#include <list>
#include <algorithm>
#include <iostream>

void BoardTest::test_black_starts() {
  CPPUNIT_ASSERT( PLAYER_1 == gameboard->get_turn() );
}

void BoardTest::test_right_root_number() {
  CPPUNIT_ASSERT( 16 == gameboard->roots.size() ); 
}

void BoardTest::test_start_position() {
    // Starting positions for PLAYER_2 (white)
    boost::shared_ptr<PointTanbo> point = gameboard->at(6, 0); 
    CPPUNIT_ASSERT( PLAYER_2 == point->color );
    point = gameboard->at(18, 0);
    CPPUNIT_ASSERT( PLAYER_2 == point->color );
    point = gameboard->at(0, 6);
    CPPUNIT_ASSERT( PLAYER_2 == point->color );
    point = gameboard->at(12, 6);
    CPPUNIT_ASSERT( PLAYER_2 == point->color );
    point = gameboard->at(6, 12);
    CPPUNIT_ASSERT( PLAYER_2 == point->color );
    point = gameboard->at(18, 12);
    CPPUNIT_ASSERT( PLAYER_2 == point->color );
    point = gameboard->at(0, 18);
    CPPUNIT_ASSERT( PLAYER_2 == point->color );
    point = gameboard->at(12, 18);
    CPPUNIT_ASSERT( PLAYER_2 == point->color );
             
    // Starting positions for PLAYER_1 (black)
    point = gameboard->at(0, 0);
    CPPUNIT_ASSERT( PLAYER_1 == point->color );
    point = gameboard->at(12, 0);
    CPPUNIT_ASSERT( PLAYER_1 == point->color );
    point = gameboard->at(6, 6);
    CPPUNIT_ASSERT( PLAYER_1 == point->color );
    point = gameboard->at(18, 6);
    CPPUNIT_ASSERT( PLAYER_1 == point->color );
    point = gameboard->at(12, 12);
    CPPUNIT_ASSERT( PLAYER_1 == point->color );
    point = gameboard->at(18, 18);
    CPPUNIT_ASSERT( PLAYER_1 == point->color );
    point = gameboard->at(6, 18);
    CPPUNIT_ASSERT( PLAYER_1 == point->color );
    point = gameboard->at(0, 12);
    CPPUNIT_ASSERT( PLAYER_1 == point->color );

   // Empty points (NOT_PLAYED)
    for(int x=0; x<19; x++) {
      for(int y=0; y<19; y++) {
       if( x%6 != 0 || y%6 != 0) {
         point = gameboard->at(x, y);
         CPPUNIT_ASSERT( NOT_PLAYED == point->color );
       }
     }
   }
}

void BoardTest::test_root_points() {
    std::vector< boost::shared_ptr<PointTanbo> > root_points = std::vector< boost::shared_ptr<PointTanbo> >();
    root_points.push_back(gameboard->at(6 ,0 ));
    root_points.push_back(gameboard->at(18,0 ));
    root_points.push_back(gameboard->at(0 ,6 ));
    root_points.push_back(gameboard->at(12,6 ));
    root_points.push_back(gameboard->at(6 ,12));
    root_points.push_back(gameboard->at(18,12));
    root_points.push_back(gameboard->at(0 ,18));
    root_points.push_back(gameboard->at(12,18));
    root_points.push_back(gameboard->at(0 ,0 ));
    root_points.push_back(gameboard->at(12,0 ));
    root_points.push_back(gameboard->at(6 ,6 ));
    root_points.push_back(gameboard->at(18,6 ));
    root_points.push_back(gameboard->at(12,12));
    root_points.push_back(gameboard->at(18,18));
    root_points.push_back(gameboard->at(6 ,18));
    root_points.push_back(gameboard->at(0 ,12));

    //Get the roots from the gameboard for iterating over
    std::vector < boost::shared_ptr<RootTanbo> > the_roots = gameboard->roots;

    for (std::vector< boost::shared_ptr<PointTanbo> >::iterator itr = root_points.begin(); itr != root_points.end(); ++itr ) {
        boost::shared_ptr<PointTanbo> point = (*itr);
        std::vector< boost::shared_ptr<PointTanbo> > found = std::vector< boost::shared_ptr<PointTanbo> >();
      
        //Find the point in the roots
        for( std::vector< boost::shared_ptr<RootTanbo> >::iterator itr = the_roots.begin(); itr != the_roots.end(); ++itr ) {
            boost::shared_ptr<RootTanbo> cur_root = (*itr);
            std::vector< boost::shared_ptr<PointTanbo> > cur_points = cur_root->points;
            
            std::vector< boost::shared_ptr<PointTanbo> >::iterator found_point, search_start;
            search_start = cur_points.begin();
            while(true) {
              found_point = std::find(search_start, cur_points.end(), point);
              if(found_point == cur_points.end()) {
                break;
              }
              found.push_back(*found_point);
              search_start = found_point + 1;
            }
        }
        
        //Should have found the point in exactly one root
        CPPUNIT_ASSERT( found.size() == 1 );
    }
}

void BoardTest::test_valid_move() {
    // Test PLAYER2 (white) starting moves
    boost::shared_ptr<PointTanbo> point = gameboard->at(0, 7);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(0, 5);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(1, 18);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(0, 17);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(5, 12);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(6, 13);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_2) );
  
    // Test PLAYER1 (black) starting moves
    point = gameboard->at(0, 1);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_1) );
    point = gameboard->at(1, 0);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_1) );
    point = gameboard->at(17, 18);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_1) );
    point = gameboard->at(18, 17);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_1) );
    point = gameboard->at(7, 6);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_1) );
    point = gameboard->at(6, 7);
    CPPUNIT_ASSERT( gameboard->is_move_valid(*point, PLAYER_1) );
}

void BoardTest::test_no_invalid_move() {  
    boost::shared_ptr<PointTanbo> point = gameboard->at(19,19);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(19,-1);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(3,4);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(4,4);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(5,4);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(9,0);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(0,10);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(10,0);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(0,18);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(6,6);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(10,18);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
    point = gameboard->at(18,10);
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_1) );
    CPPUNIT_ASSERT( ! gameboard->is_move_valid(*point, PLAYER_2) );
}

  void BoardTest::test_root_liberties() {
    std::vector< boost::shared_ptr<PointTanbo> > root_points = std::vector< boost::shared_ptr<PointTanbo> >();
    root_points.push_back(gameboard->at(6 ,0 ));
    root_points.push_back(gameboard->at(18,0 ));
    root_points.push_back(gameboard->at(0 ,6 ));
    root_points.push_back(gameboard->at(12,6 ));
    root_points.push_back(gameboard->at(6 ,12));
    root_points.push_back(gameboard->at(18,12));
    root_points.push_back(gameboard->at(0 ,18));
    root_points.push_back(gameboard->at(12,18));
    root_points.push_back(gameboard->at(0 ,0 ));
    root_points.push_back(gameboard->at(12,0 ));
    root_points.push_back(gameboard->at(6 ,6 ));
    root_points.push_back(gameboard->at(18,6 ));
    root_points.push_back(gameboard->at(12,12));
    root_points.push_back(gameboard->at(18,18));
    root_points.push_back(gameboard->at(6 ,18));
    root_points.push_back(gameboard->at(0 ,12));
    
    //Get the roots from the gameboard for iterating over
    std::vector < boost::shared_ptr<RootTanbo> > the_roots = gameboard->roots;

    for (std::vector< boost::shared_ptr<PointTanbo> >::iterator itr = root_points.begin(); itr != root_points.end(); ++itr ) {
        boost::shared_ptr<PointTanbo> point = (*itr);
        std::vector< boost::shared_ptr<PointTanbo> > found = std::vector< boost::shared_ptr<PointTanbo> >();
      
        //Find the point in the roots
        for( std::vector< boost::shared_ptr<RootTanbo> >::iterator itr = the_roots.begin(); itr != the_roots.end(); ++itr ) {
            boost::shared_ptr<RootTanbo> cur_root = (*itr);
            std::vector< boost::shared_ptr<PointTanbo> > cur_points = cur_root->points;
            
            std::vector< boost::shared_ptr<PointTanbo> >::iterator found_point, search_start;
            search_start = cur_points.begin();
            while(true) {
              found_point = std::find(search_start, cur_points.end(), point);
              if(found_point == cur_points.end()) {
                break;
              }
              search_start = found_point + 1;
              
              //We found the root we're looking for (contains point)
              PointTanboVecPtr bnd_neighbors = point->bounded_neighbors();
              for( std::vector< boost::shared_ptr<PointTanbo> >::iterator itr = bnd_neighbors->begin(); itr != bnd_neighbors->end(); ++itr ) {
                boost::shared_ptr<PointTanbo> cur_nbor = (*itr);
                bool lib_found = false;
                std::list< boost::shared_ptr<PointTanbo> > root_libs = cur_root->liberties;
                
                // std::cout << "****" << std::endl;
                // cur_root->print();
                // // for(std::list< boost::shared_ptr<PointTanbo> >::iterator itr_1 = root_libs.begin(); itr_1 != root_libs.end(); ++itr_1) {
                // //   (*(*itr_1)).print();
                // // }
                // std::cout << "****" << std::endl;
                
                std::list< boost::shared_ptr<PointTanbo> >::iterator found_point_lib, search_start_lib;
                search_start_lib = root_libs.begin();
                while(true) {
                  found_point_lib = std::find(search_start_lib, root_libs.end(), cur_nbor);
                  // find() returns the end of list if nothing is found
                  if(found_point_lib == root_libs.end()) {
                    break;
                  } else {
                    lib_found = true;
                    break;
                  }
                }
                
                //Assert that each of the neighbors of the point of the root was returned as a liberty of that root
                CPPUNIT_ASSERT( lib_found );
              }
            }
        }
    }
  }
